numba/tests/test_dictimpl.py

```diff
@@ -181,6 +186,80 @@ def __next__(self):
     next = __next__    # needed for py2 only
 
 
+############################################################
+# Supporting code for TestDictImpl.test_parametrized_types
+############################################################
+class Parametrized(tuple):
+    def __init__(self, tup):
+        assert all(isinstance(v, str) for v in tup)
+
+
+class ParametrisedType(types.Type):
+    """this is essentially UniTuple(unicode_type, n)
+    BUT type name is the same for all n"""
+    def __init__(self, value):
+        super(ParametrisedType, self).__init__('ParametrisedType')
+        self.dtype = types.unicode_type
+        self.n = len(value)
+
+    @property
+    def key(self):
+        return self.n
+
+    def __len__(self):
+        return self.n
+
+
+register_model(ParametrisedType)(UniTupleModel)
+
+
+@typeof_impl.register(Parametrized)
+def typeof_unit(val, c):
+    return ParametrisedType(val)
+
+
+@unbox(ParametrisedType)
+def unbox_parametrized(typ, obj, context):
+    return context.unbox(types.UniTuple(typ.dtype, len(typ)), obj)
+
+
+@generated_jit
+def dict_vs_cache_vs_parametrized(v):
+    typ = v
+
+    def objmode_vs_cache_vs_parametrized_impl(v):
+        # typed.List shows same behaviour after fix for #6397
+        d = typed.Dict.empty(types.unicode_type, typ)
+        d['data'] = v
+
+    return objmode_vs_cache_vs_parametrized_impl
+
+
+def dict_vs_cache_vs_parametrized_worker():
+    """crashes when run more than once
+    (ie when compiled function is loaded from cache)"""
+    @jit(nopython=True, cache=True)
+    def get_unit_system_data(x, y):
+        dict_vs_cache_vs_parametrized(x)
+        dict_vs_cache_vs_parametrized(y)
+
+    x, y = Parametrized(('horst', 'hanz')), Parametrized(('horst',))
+
+    for ii in range(50):  # <- somtimes works a few times
+        assert get_unit_system_data(x, y) is None
+
+
+def run_test_parametrized_types_in_subprocess():
+    env = os.environ.copy()
+    subproc = subprocess.Popen(
+        [sys.executable, '-c',
+         'import numba.tests.test_dictimpl as test_mod\n' +
+         'test_mod.dict_vs_cache_vs_parametrized_worker()'],
+        env=env)
+    subproc.wait()
+    return subproc.returncode
```

Comment:
I'm trying to understand why this is in a subprocess, is it with view of isolating a potential segfault? If the PR is fixing something that has a tendency to segfault, I'd rather it segfaults the test suite as a demonstration that it's not working than survive via isolation and report it, else we'd end up with all tests in subprocesses just-in-case?

Reply:
I have not found a reproducer w/o subprocess, yet. I've tried and `Dispatcher._make_finalizer(()(); Dispatcher._reset_overloads(); jit(...)` to force re-loading from cache (which seems to do the job) but it still does not crash unless the cached function is used in another process. Is there anything more I can try and reset/delete in a single process?

Reply:
After a little more experimentation I came-up with the below. Does the way I am resetting the dispatchers and the target context look resonable?
```
    def test_parametrized_types(self):
        """https://github.com/numba/numba/issues/6401"""

        register_model(ParametrizedType)(UniTupleModel)

        @typeof_impl.register(Parametrized)
        def typeof_unit(val, c):
            return ParametrizedType(val)

        @unbox(ParametrizedType)
        def unbox_parametrized(typ, obj, context):
            return context.unbox(types.UniTuple(typ.dtype, len(typ)), obj)

        @generated_jit
        def dict_vs_cache_vs_parametrized(v):
            typ = v

            def objmode_vs_cache_vs_parametrized_impl(v):
                # typed.List shows same behaviour after fix for #6397
                d = typed.Dict.empty(types.unicode_type, typ)
                d['data'] = v

            return objmode_vs_cache_vs_parametrized_impl

        @jit(nopython=True, cache=True)
        def set_parametrized_data(x, y):
            # Has had a tendency to segfault when the compiled function
            # was loaded from cache in a different process than the one
            # it was originally compiled in.
            # The new process is simulated below by resetting the dispatchers
            # and the target context
            dict_vs_cache_vs_parametrized(x)
            dict_vs_cache_vs_parametrized(y)

        x, y = Parametrized(('a', 'b')), Parametrized(('a',))
        set_parametrized_data(x, y)

        # reset dispatchers and targetctx to force re-load from cache as a new process would jit
        set_parametrized_data._make_finalizer()()
        set_parametrized_data._reset_overloads()
        set_parametrized_data.targetctx.init()

        for ii in range(50):  # <- somtimes works a few times
            assert set_parametrized_data(x, y) is None
```